web/documentserver-example/csharp/FormatManager.cs

```diff
@@ -0,0 +1,188 @@
+        {
+            return All()
+                .Where(format => format.Actions.Contains("fill"))
+                .ToList();
+        }
+
+        public static List<string> ViewableExtensions()
+        {
+            return Viewable()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Viewable()
+        {
+            return All()
+                .Where(format => format.Actions.Contains("view"))
+                .ToList();
+        }
+
+        public static List<string> EditableExtensions()
+        {
+            return Editable()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Editable()
+        {
+            return All()
+                .Where(format => format.Actions.Contains("edit") || format.Actions.Contains("lossy-edit"))
+                .ToList();
+        }
+
+        public static List<string> ConvertibleExtensions()
+        {
+            return Convertible()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Convertible()
+        {
+            return All()
+                .Where(format => (format.Type == "cell" && format.Convert.Contains("xlsx"))
+                                || (format.Type == "slide" && format.Convert.Contains("pptx"))
+                                || (format.Type == "word" && format.Convert.Contains("docx")))
+                .ToList();
+        }
+
+        public static List<string> SpreadsheetExtensions()
+        {
+            return Spreadsheets()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Spreadsheets()
+        {
+            return All()
+                .Where(format => format.Type == "cell")
+                .ToList();
+        }
+
+        public static List<string> PresentationExtensions()
+        {
+            return Presentations()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Presentations()
+        {
+            return All()
+                .Where(format => format.Type == "slide")
+                .ToList();
+        }
+
+        public static List<string> DocumentExtensions()
+        {
+            return Documents()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> Documents()
+        {
+            return All()
+                .Where(format => format.Type == "word")
+                .ToList();
+        }
+
+        public static List<string> AllExtensions()
+        {
+            return All()
+                .Select(format => format.Extension())
+                .ToList();
+        }
+
+        public static List<Format> All()
```

Comment:
Maybe initialize the formats array when creating an instance of the class?

Reply:
> Maybe initialize the formats array when creating an instance of the class?

I added format caching